src/tests/ftest/cart/dual_provider_common.h

```diff
@@ -0,0 +1,188 @@
+/*
+ * (C) Copyright 2019-2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+#ifndef __DUAL_PROVIDER_COMMON_H__
+#define __DUAL_PROVIDER_COMMON_H__
+static int do_shutdown;
+static int g_my_rank;
+
+
+static void
+exit_on_line(int line)
+{
+	printf("Failed on line %d\n", line);
+	exit(0);
+}
+
+#define error_exit() exit_on_line(__LINE__)
+
+
+#define MY_BASE 0x010000000
+#define MY_VER  0
+
+#define NUM_PRIMARY_CTX_MAX 8
+#define NUM_SECONDARY_CTX_MAX 8
+
+#define SERVER_GROUP_NAME "dual_provider_group"
+
+#define RPC_DECLARE(name)					\
+	CRT_RPC_DECLARE(name, CRT_ISEQ_##name, CRT_OSEQ_##name)	\
+	CRT_RPC_DEFINE(name, CRT_ISEQ_##name, CRT_OSEQ_##name)
+
+enum {
+	RPC_PING = CRT_PROTO_OPC(MY_BASE, MY_VER, 0),
+	RPC_SHUTDOWN
+} rpc_id_t;
+
+#define CRT_ISEQ_RPC_PING	/* input fields */		 \
+	((crt_bulk_t)		(bulk_hdl1)		CRT_VAR) \
+	((crt_bulk_t)		(bulk_hdl2)		CRT_VAR) \
+	((uint32_t)		(size1)			CRT_VAR) \
+	((uint32_t)		(size2)			CRT_VAR)
+
+#define CRT_OSEQ_RPC_PING	/* output fields */		 \
+	((crt_bulk_t)		(ret_bulk)		CRT_VAR) \
+	((int32_t)		(rc)			CRT_VAR)
+
+#define CRT_ISEQ_RPC_SHUTDOWN	/* input fields */		 \
+	((uint32_t)		(field)			CRT_VAR)
+
+#define CRT_OSEQ_RPC_SHUTDOWN	/* output fields */		 \
+	((uint32_t)		(field)			CRT_VAR)
+
+
+static int handler_ping(crt_rpc_t *rpc);
+static int handler_shutdown(crt_rpc_t *rpc);
+
+RPC_DECLARE(RPC_PING);
+RPC_DECLARE(RPC_SHUTDOWN);
+
+struct crt_proto_rpc_format my_proto_rpc_fmt[] = {
+	{
+		.prf_flags	= 0,
+		.prf_req_fmt	= &CQF_RPC_PING,
+		.prf_hdlr	= (void *)handler_ping,
+		.prf_co_ops	= NULL,
+	}, {
+		.prf_flags	= 0,
+		.prf_req_fmt	= &CQF_RPC_SHUTDOWN,
+		.prf_hdlr	= (void *)handler_shutdown,
+		.prf_co_ops	= NULL,
+	}
+};
+
+struct crt_proto_format my_proto_fmt = {
+	.cpf_name = "my-proto",
+	.cpf_ver = MY_VER,
+	.cpf_count = ARRAY_SIZE(my_proto_rpc_fmt),
+	.cpf_prf = &my_proto_rpc_fmt[0],
+	.cpf_base = MY_BASE,
+};
+
+static int
+bulk_transfer_done_cb(const struct crt_bulk_cb_info *info)
+{
+	void	*buff;
+	int	rc;
+
+	DBG_PRINT("Bulk transfer failed with rc=%d\n", info->bci_rc);
+	if (info->bci_rc != 0) {
+		error_exit();
+	}
+
+	DBG_PRINT("Bulk transfer done\n");
+
+	rc = crt_reply_send(info->bci_bulk_desc->bd_rpc);
+	if (rc != 0) {
+		D_ERROR("Failed to send response\n");
+		error_exit();
+	}
+
+	crt_bulk_free(info->bci_bulk_desc->bd_local_hdl);
+
+	buff = info->bci_arg;
+	D_FREE(buff);
+
+	RPC_PUB_DECREF(info->bci_bulk_desc->bd_rpc);
+
+	return 0;
+}
+
+static int
+handler_ping(crt_rpc_t *rpc)
+{
+	struct RPC_PING_in	*input;
+	struct RPC_PING_out	*output;
+	crt_context_t		*ctx;
+	int			rc = 0;
+
+	input = crt_req_get(rpc);
+	output = crt_reply_get(rpc);
+
+	output->rc = 0;
+	DBG_PRINT("Sizes: %d %d\n", input->size1, input->size2);
+
+	ctx = rpc->cr_ctx;
+
+	DBG_PRINT("RPC arived on a %s context\n",
+		  crt_context_is_primary(ctx) ? "primary" : "secondary");
+#if 1
```

Comment:
Could this #if / #endif pair be removed now?

Reply:
yes indeed